governance/default_test.go

```diff
@@ -560,3 +563,201 @@ func TestWriteGovernance_idxCache(t *testing.T) {
+	gov := getGovernance()
+	gov.nodeAddress.Store(council[len(council)-1])
+
+	votes := make([]GovernanceVote, 0)
+	tally := make([]GovernanceTallyItem, 0)
+
+	proposer := council[0]
+	self := council[len(council)-1]
+	header := &types.Header{}
+
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test for "none" mode
+	header.Number = blockCounter.Add(blockCounter, common.Big1)
+	header.BlockScore = common.Big1
+	gov.AddVote("governance.unitprice", uint64(22000))
+	header.Vote = gov.GetEncodedVote(proposer, blockCounter.Uint64())
+
+	gov.HandleGovernanceVote(valSet, votes, tally, header, proposer, self)
+
+	if _, ok := gov.changeSet.items["governance.unitprice"]; !ok {
+		t.Errorf("Vote had to be applied but it wasn't")
+	}
+
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test for wrong key
+	header.Number = blockCounter.Add(blockCounter, common.Big1)
+	gov.AddVote("governance.non-existent", true)
+	header.Vote = gov.GetEncodedVote(proposer, blockCounter.Uint64())
+
+	gov.HandleGovernanceVote(valSet, votes, tally, header, proposer, self)
+
+	if _, ok := gov.changeSet.items["governance.non-existent"]; ok {
+		t.Errorf("Vote shouln't be applied but it was")
+	}
+	gov.voteMap.Clear()
+
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test removing a validator
+	header.Number = blockCounter.Add(blockCounter, common.Big1)
+	gov.AddVote("governance.removevalidator", council[1].String())
+	header.Vote = gov.GetEncodedVote(proposer, blockCounter.Uint64())
+
+	gov.HandleGovernanceVote(valSet, votes, tally, header, proposer, self)
+	if i, _ := valSet.GetByAddress(council[1]); i != -1 {
+		t.Errorf("Validator removal failed, %d validators remains", valSet.Size())
+	}
+	gov.voteMap.Clear()
+
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test adding a validator
+	header.Number = blockCounter.Add(blockCounter, common.Big1)
+	gov.AddVote("governance.addvalidator", council[1].String())
+	header.Vote = gov.GetEncodedVote(proposer, blockCounter.Uint64())
+
+	gov.HandleGovernanceVote(valSet, votes, tally, header, proposer, self)
+	if i, _ := valSet.GetByAddress(council[1]); i == -1 {
+		t.Errorf("Validator addition failed, %d validators remains", valSet.Size())
+	}
+	gov.voteMap.Clear()
+}
+
+func TestGovernance_HandleGovernanceVote_Ballot_mode(t *testing.T) {
+	// Create ValidatorSet
+	council := getTestCouncil()
+	rewards := getTestRewards()
+
+	blockCounter := common.Big0
+	var valSet istanbul.ValidatorSet
+	valSet = validator.NewWeightedCouncil(council, rewards, getTestVotingPowers(len(council)), nil, istanbul.WeightedRandom, 21, 0, 0, nil)
+
+	config := getTestConfig()
+	config.Governance.GovernanceMode = "ballot"
```

Comment:
@ehnuje Actually there is a constant `governance.GovernanceMode_Ballot`. But it have to be written in string in `genesis.json`. The above code is to prepare `chainConfig` which holds original values of genesis.json and it will be mapped to the constant using `GovernanceModeMap`.

Reply:
@andybclee I think we can avoid misuse of string literals by using dedicated type for `GovernanceMode`, like below. How do you think?

```
type GovernanceMode string
const GovernanceModeBallot GovernanceMode = "ballot"
```

Reply:
@ehnuje I think it would be good to use a constant for adding more test but I'm not sure adding a type is necessary because it will require another typecasting in setting the value.  I added a constant and please take another look. =)

Reply:
Oh, what I mean is not just for test, but for general usage of `GovernanceMode`. I think there seems to be no problem if we declare a new type from `string`.

Reply:
In current code base, after the initialization, following integer constants are used for `GovernanceMode`. String values are used in two places, in the `genesis.json` and RPC API argument.

```
params.params.GovernanceMode_None
params.GovernanceMode_Single
params.GovernanceMode_Ballot
```

I think if we need a new type for `GovernanceMode`, it will be better to apply it to the above constants because those values are just integer at the moment.

But in my opinion, it is another refactoring work and needs to be separated from this PR.